docs/RFC/draft-paragonie-paseto-rfc-00.txt

```diff
@@ -0,0 +1,1680 @@
+
+
+
+
+
+Arciszewski & Haussmann Expires October 21, 2018               [Page 14]
+
+Internet-Draft                                                April 2018
+
+
+   4.  Pack "h", "m", and "f" together using PAE (see Section 2.2).
+       We'll call this "m2".
+
+   5.  Use Ed25519 to verify that the signature is valid for the
+       message: (See below for pseudocode.)
+
+   6.  If the signature is valid, return "m".  Otherwise, throw an
+       exception.
+
+                   valid = crypto_sign_verify_detached(
+                       signature = s,
+                       message = m2,
+                       public_key = pk
+                   );
+
+         Step 5: Validating the Ed25519 signature using libsodium.
+
+6.  Payload Processing
+
+   All PASETO payloads MUST be a JSON-encoded object represented as a
+   UTF-8 encoded string.  The topmost JSON object should be an object,
+   map, or associative array (depending on the language of choice), not
+   a flat array.
+
+      *Valid*:
+
+
+
+      *  "{"foo":"bar"}"
+
+      *  "{"foo":"bar","baz":12345,"678":["a","b","c"]}"
+
+      *Invalid*:
+
+
+
+      *  "[{"foo":"bar"}]"
+
+      *  "["foo"]"
+
+      *  "{0: "test"}"
+
+   If non-UTF-8 character sets are desired for some fields, implementors
+   are encouraged to use Base64url [1] encoding to preserve the original
+   intended binary data, but still use UTF-8 for the actual payloads.
+
+
+
+
+
+
+Arciszewski & Haussmann Expires October 21, 2018               [Page 15]
+
+Internet-Draft                                                April 2018
+
+
+6.1.  Registered Claims
+
+   The following keys are reserved for use within PASETO.  Users SHOULD
+   NOT write arbitrary/invalid data to any keys in a top-level PASETO in
+   the list below:
+
+    +-----+------------+--------+-------------------------------------+
+    | Key |    Name    |  Type  |               Example               |
+    +-----+------------+--------+-------------------------------------+
+    | iss |   Issuer   | string |       {"iss":"paragonie.com"}       |
+    | sub |  Subject   | string |            {"sub":"test"}           |
+    | aud |  Audience  | string |       {"aud":"pie-hosted.com"}      |
+    | exp | Expiration | DtTime | {"exp":"2039-01-01T00:00:00+00:00"} |
+    | nbf | Not Before | DtTime | {"nbf":"2038-04-01T00:00:00+00:00"} |
+    | iat | Issued At  | DtTime | {"iat":"2038-03-17T00:00:00+00:00"} |
+    | jti |  Token ID  | string |  {"jti":"87IFSGFgPNtQNNuw0AtuLttP"} |
+    +-----+------------+--------+-------------------------------------+
+
+   In the table above, DtTime means an ISO 8601 compliant DateTime
+   string.
+
+   Any other claims can be freely used.  These keys are only reserved in
+   the top-level JSON object.
+
+   The keys in the above table are case-sensitive.
+
+   Implementors SHOULD provide some means to discourage setting invalid/
+   arbitrary data to these reserved claims.
```

Comment:
I'm unclear what this means - do you mean library writers or any users? What would be an example of invalid/arbitrary data that wouldn't match the given types?